UOP1_Project/Assets/Scripts/DesignPatterns/StateMachine/StateMachine.cs

```diff
@@ -0,0 +1,64 @@
+﻿using System;
+using System.Collections.Generic;
+using UnityEngine;
+
+public class StateMachine
+{
+	private State currentState;
+
+	private readonly List<Transition> transitionFromAnyState = new List<Transition>();
+
+	public void Tick()
+	{
+		Transition transition = GetTransitionIfAvailable(currentState);
+		if (transition != null)
+		{
+			SetState(transition.To);
+		}
+
+		currentState?.Tick();
+	}
+
+	public void SetState(State state)
+	{
+		if (state == currentState)
+		{
+			return;
+		}
+
+		currentState?.OnExit();
+		currentState = state;
+
+		currentState.OnEnter();
+	}
+
+	public void AddAnyTransition(State state, Func<bool> predicate)
+	{
+		transitionFromAnyState.Add(new Transition(state, predicate));
+	}
+
+	private Transition GetTransitionIfAvailable(State currentState)
+	{
+		Transition transitionIfAny = GetNextTransitionFromList(transitionFromAnyState);
+
+		if (transitionIfAny == null || transitionIfAny.To == currentState)
+		{
+			transitionIfAny = GetNextTransitionFromList(this.currentState.Transitions);
+		}
+
+		return transitionIfAny;
+	}
+
+	private static Transition GetNextTransitionFromList(List<Transition> transitions)
+	{
+		foreach (Transition transition in transitions)
+		{
+			if (transition.Condition())
+			{
+				return transition;
```

Comment:
Problem with two (or more) transitionFromAnyState : If the first is Ok but transitionIfAny.To = currentState, it is returned and the other not even tested (or used if they are valid).